### ISSUE TITLE
How can I do binary classification?

### ISSUE THREAD

**Issue by johnny5550822** [vlfeat/matconvnet#47]
I saw that Matconvnet allow me to choose errortype as 'binary'. Is that the only parameter tht I need to change if I want to do binary classification (my label is 1 or 2).

**Comment by lenck** [comment]
Hi, based on the `cnn_train.m` example for training, you need to have labels +-1. But as it is an example training code, there is nothing easier to change the line 274 in [cnn_train.m](https://github.com/vlfeat/matconvnet/blob/21320ac921754efaeb9a99d481b2a907d9592376/examples/cnn_train.m) to:
```
 error = bsxfun(@eq, predictions, labels);
```
And then you should be able to use 1,2 labels. It's just that it is sort of a common practice to have binary task with labels +-1...

**Comment by johnny5550822** [comment]
I see. Great. Thanks!

**Comment by lenck** [comment]
Hmm, I just realised that I got it slightly wrong :)
With the binary classification, the network actually has a single output - that's why you multiply with the label directly...
But I just found, that actually having 2-class softmax is [the same as logistic regression](http://ufldl.stanford.edu/wiki/index.php/Softmax_Regression#Relationship_to_Logistic_Regression) so in order to use what I've proposed, you will need to have `softmaxloss` layer at the end with two outputs. But that might have actually been your case, so I hope that it worked :)

**Comment by johnny5550822** [comment]
Hmm, so should I still use -1/+1?

**Comment by lenck** [comment]
I think the easiest would be to use the softmax with two outputs and two labels 1 and 2, then you just need to remove the top5 error evaluation in updateError in `cnn_train` (as you will have just 2 labels)...

It's actually worth to fix this - to be able to set the `n` for the error measures...

**Comment by johnny5550822** [comment]
Great. Or actually I have data only with 2 labels (1 or 2) and I still set my output as having 5 distinct labels to train the network. I think this will still be ok, am I right?

**Comment by lenck** [comment]
Hmm, I'm not sure. At least the last layer would be slower, because you would have more parameters to learn (in this case to learn that they should be zero) - so you would deliberately make your problem more complicated.
But it's really just about commmenting out lines {271, 272} in `cnn_train.m` :)

**Comment by amar2015** [comment]
hi 
thanks a lot for offering that great tool box. I am trying to run cnn_mint but I have this error:

learning rate changed (0.000000 --> 0.001000): resetting momentum
training: epoch 01: processing batch   1 of 600 ...Error using  .* 
Array dimensions must match for binary array op.

Error in vl_nnsoftmax (line 30)
Y = Y .* bsxfun(@minus, dzdY, sum(dzdY .* Y, 3)) ;

Error in vl_simplenn (line 211)
        res(i).dzdx = vl_nnsoftmax(res(i).x, res(i+1).dzdx) ;

Error in cnn_train (line 140)
    res = vl_simplenn(net, im, one, res, ...

Error in cnn_mnist (line 75)
[net, info] = cnn_train(net, imdb, @getBatch, ...
 
I really appreciate it if you help to solve that error because I am trying to use the library to classify my own data.

**Comment by johnny5550822** [comment]
ok, great! Thanks.

**Comment by johnny5550822** [comment]
I comment the {271,272} lines and keep label as 1/2, the output of the error plot look weird to me...

![screenshot from 2015-01-28 14 41 11](https://cloud.githubusercontent.com/assets/1665063/5948703/d35595cc-a6fb-11e4-92b7-9b9b196ab9f7.png)

**Comment by cheer37** [comment]
have you solved your problem?
I did binary classification with binary train type and 1/-1 labels.
But result doesn't make sense and graph is like you.
So, i did train with multiclass type and 1/2 labels.
But it does not converge.
How did you deal with binary classification with matconvnet?

**Comment by anewlearner** [comment]
@lenck 
I did just as what you said,but things seems to be wrong.
In cifar, I changed the 10 in penultimate layers to be 2, and have the labels of 1,2. but the error didn't change in epochs. It always be obj:0.693 top1err:0.501.
I also tried to use the lables of -1 and 1. And change the error to 'error = bsxfun(@eq, predictions, labels);'  But I just get a large error such as 150.
Look forward to your reply!
Thanks!
